File: software/examples/qollective_mcp_websocket_holodeck/holodeck-desktop/src/types/holodeck.ts
// ABOUTME: TypeScript type definitions matching Rust shared-types business models
// ABOUTME: Provides type safety for React components interacting with holodeck data

export enum HolodeckStoryType {
  Adventure = 'Adventure',
  Mystery = 'Mystery',
  Drama = 'Drama',
  Comedy = 'Comedy',
  Historical = 'Historical',
  SciFi = 'SciFi',
  Fantasy = 'Fantasy',
  Educational = 'Educational',
}

export enum SafetyLevel {
  Training = 'Training',
  Standard = 'Standard',
  Reduced = 'Reduced',
  Disabled = 'Disabled',
}

export enum SessionStatus {
  Active = 'Active',
  Paused = 'Paused',
  Completed = 'Completed',
  Aborted = 'Aborted',
  SafetyHalt = 'SafetyHalt',
}

export interface Position3D {
  x: number;
  y: number;
  z: number;
}

export interface EnvironmentalControls {
  temperatureCelsius: number;
  humidityPercent: number;
  atmosphericPressure: number;
  oxygenLevel: number;
  windSimulation: boolean;
  weatherEffects: boolean;
}

export interface HolodeckConfig {
  safetyLevel: SafetyLevel | string;
  maxParticipants: number;
  durationMinutes?: number;
  autoSaveEnabled: boolean;
  voiceRecognition: boolean;
  hapticFeedback: boolean;
  replicatorAccess: boolean;
  transporterIntegration: boolean;
  environmentalControls: EnvironmentalControls;
}

export interface Scene {
  id: string;
  name: string;
  description: string;
  environmentId: string;
  charactersPresent: string[];
  props: SceneProp[];
  backgroundAudio?: string;
  lightingConfig: LightingConfig;
  physicsSettings: PhysicsSettings;
}

export interface SceneProp {
  id: string;
  name: string;
  description: string;
  position: Position3D;
  interactive: boolean;
  physicsEnabled: boolean;
}

export interface LightingConfig {
  ambientLight: number;
  directionalLights: DirectionalLight[];
  mood: LightingMood;
}

export interface DirectionalLight {
  direction: Position3D;
  intensity: number;
  color: Color;
}

export interface Color {
  r: number;
  g: number;
  b: number;
  a: number;
}

export enum LightingMood {
  Bright = 'Bright',
  Dim = 'Dim',
  Dramatic = 'Dramatic',
  Cozy = 'Cozy',
  Mysterious = 'Mysterious',
  Tense = 'Tense',
}

export interface PhysicsSettings {
  gravityEnabled: boolean;
  gravityStrength: number;
  collisionDetection: boolean;
  realTimePhysics: boolean;
}

export interface Holodeck {
  id: string;
  name: string;
  topic: string;
  storyType: HolodeckStoryType;
  participants: Character[];
  currentScene?: Scene;
  configuration: HolodeckConfig;
  createdAt: Date;
  updatedAt: Date;
}

export interface Character {
  id: string;
  name: string;
  characterType: CharacterType;
  personality: PersonalityTraits;
  voiceConfig: VoiceConfig;
  appearance: CharacterAppearance;
  knowledgeDomains: KnowledgeDomain[];
  relationships: CharacterRelationship[];
  currentMood: Mood;
  position?: Position3D;
  status: CharacterStatus;
}

export enum CharacterType {
  Captain = 'Captain',
  FirstOfficer = 'FirstOfficer',
  Android = 'Android',
  ChiefEngineer = 'ChiefEngineer',
  ChiefMedicalOfficer = 'ChiefMedicalOfficer',
  Counselor = 'Counselor',
  ChiefOfSecurity = 'ChiefOfSecurity',
  NavigationOfficer = 'NavigationOfficer',
  BarTender = 'BarTender',
  Civilian = 'Civilian',
  HistoricalFigure = 'HistoricalFigure',
  Custom = 'Custom',
}

export interface PersonalityTraits {
  logical: number; // 0.0 - 1.0
  emotional: number; // 0.0 - 1.0
  authoritative: number; // 0.0 - 1.0
  diplomatic: number; // 0.0 - 1.0
  curious: number; // 0.0 - 1.0
  cautious: number; // 0.0 - 1.0
  humorTendency: number; // 0.0 - 1.0
  leadershipStyle: LeadershipStyle;
  communicationStyle: CommunicationStyle;
}

export enum LeadershipStyle {
  Commanding = 'Commanding',
  Collaborative = 'Collaborative',
  Technical = 'Technical',
  Supportive = 'Supportive',
  Analytical = 'Analytical',
}

export enum CommunicationStyle {
  Direct = 'Direct',
  Diplomatic = 'Diplomatic',
  Technical = 'Technical',
  Philosophical = 'Philosophical',
  Empathetic = 'Empathetic',
  Formal = 'Formal',
}

export interface VoiceConfig {
  voiceActor: string;
  speechPatterns: SpeechPattern[];
  commonPhrases: string[];
  accent?: string;
  toneModulation: ToneModulation;
}

export interface SpeechPattern {
  patternType: SpeechPatternType;
  frequency: number; // 0.0 - 1.0
  examples: string[];
}

export enum SpeechPatternType {
  CatchPhrase = 'CatchPhrase',
  TechnicalJargon = 'TechnicalJargon',
  PhilosophicalReference = 'PhilosophicalReference',
  MilitaryProtocol = 'MilitaryProtocol',
  ScientificExplanation = 'ScientificExplanation',
  EmotionalExpression = 'EmotionalExpression',
  LogicalAnalysis = 'LogicalAnalysis',
}

export interface ToneModulation {
  basePitch: number;
  emotionalRange: number;
  formalityLevel: number;
  speechSpeed: number;
}

export interface CharacterAppearance {
  species: Species;
  heightCm: number;
  uniformColor: UniformColor;
  distinctiveFeatures: string[];
  holographicFidelity: FidelityLevel;
}

export enum Species {
  Human = 'Human',
  Android = 'Android',
  Klingon = 'Klingon',
  Vulcan = 'Vulcan',
  Betazoid = 'Betazoid',
  ElAurian = 'ElAurian',
  Bolian = 'Bolian',
}

export enum UniformColor {
  Command = 'Command',
  Sciences = 'Sciences',
  Operations = 'Operations',
  Civilian = 'Civilian',
}

export enum FidelityLevel {
  Basic = 'Basic',
  Standard = 'Standard',
  HighDefinition = 'HighDefinition',
  Photorealistic = 'Photorealistic',
}

export enum KnowledgeDomain {
  Starfleet = 'Starfleet',
  Engineering = 'Engineering',
  Medical = 'Medical',
  Psychology = 'Psychology',
  Physics = 'Physics',
  Diplomacy = 'Diplomacy',
  MilitaryTactics = 'MilitaryTactics',
  Philosophy = 'Philosophy',
  History = 'History',
  AlienCultures = 'AlienCultures',
  Technology = 'Technology',
  Navigation = 'Navigation',
}

export interface CharacterRelationship {
  otherCharacterId: string;
  relationshipType: RelationshipType;
  strength: number; // 0.0 - 1.0
  trustLevel: number; // 0.0 - 1.0
}

export enum RelationshipType {
  Superior = 'Superior',
  Subordinate = 'Subordinate',
  Colleague = 'Colleague',
  Friend = 'Friend',
  Mentor = 'Mentor',
  Rival = 'Rival',
  Family = 'Family',
  Romantic = 'Romantic',
}

export enum Mood {
  Calm = 'Calm',
  Excited = 'Excited',
  Concerned = 'Concerned',
  Focused = 'Focused',
  Curious = 'Curious',
  Amused = 'Amused',
  Serious = 'Serious',
  Contemplative = 'Contemplative',
}

export enum CharacterStatus {
  Active = 'Active',
  Inactive = 'Inactive',
  Busy = 'Busy',
  AwayFromStation = 'AwayFromStation',
  InMeeting = 'InMeeting',
  EmergencyProtocol = 'EmergencyProtocol',
}

export interface StoryTemplate {
  id: string;
  holodeckId: string;
  title: string;
  topic: string;
  scenes: SceneTemplate[];
  storyGraph: StoryGraph;
  metadata: StoryMetadata;
  createdAt: Date;
}

export interface SceneTemplate {
  id: string;
  sequenceNumber: number;
  title: string;
  description: string;
  environmentContext: string;
  charactersPresent: Character[];
  decisionPoints: DecisionPoint[];
  safetyRating: SafetyLevel | string;
}

export interface StoryGraph {
  nodes: { [key: string]: GraphNode };
  edges: GraphEdge[];
  startNode: string;
  endNodes: string[];
}

export interface GraphNode {
  id: string;
  sceneId: string;
  nodeType: NodeType;
  position: [number, number];
}

export enum NodeType {
  Start = 'Start',
  Scene = 'Scene',
  Decision = 'Decision',
  End = 'End',
}

export interface GraphEdge {
  from: string;
  to: string;
  condition?: string;
}

export interface DecisionPoint {
  id: string;
  description: string;
  decisionType: DecisionType;
  consequences: Consequence[];
  nextSceneId?: string;
}

export enum DecisionType {
  Continue = 'Continue',
  BranchLeft = 'BranchLeft',
  BranchRight = 'BranchRight',
  GoBack = 'GoBack',
  InteractCharacter = 'InteractCharacter',
  ExamineEnvironment = 'ExamineEnvironment',
}

export interface Consequence {
  type: string;
  description: string;
  impact: number; // -1.0 to 1.0
}

export interface StoryMetadata {
  generatedAt: Date;
  complexity: number;
  estimatedDuration: number;
  language: string;
  wordsPerScene: string;
}

export interface StoryBook {
  id: string;
  templateId: string;
  holodeckId: string;
  playerName: string;
  sessionName: string;
  playedScenes: PlayedScene[];
  currentPosition: Position3D;
  playerDecisions: PlayerDecision[];
  sessionStatistics: SessionStatistics;
  status: SessionStatus;
  startedAt: Date;
  lastPlayed: Date;
  completedAt?: Date;
}

export interface PlayedScene {
  id: string;
  title: string;
  description: string;
  characterInteractions: string[];
  playerDecisions: string[];
  completedAt: Date;
}

export interface PlayerDecision {
  id: string;
  sceneId: string;
  decision: string;
  timestamp: Date;
  responseTimeMs: number;
}

export interface SessionStatistics {
  totalScenesPlayed: number;
  totalDecisionsMade: number;
  averageResponseTimeMs: number;
  charactersInteractedWith: string[];
  storyPathsExplored: string[];
  safetyInterventions: number;
  achievementUnlocked: Achievement[];
}

export interface Achievement {
  id: string;
  name: string;
  description: string;
  unlockedAt: Date;
  iconUrl?: string;
}

// UI-specific types
export interface WelcomeData {
  playerName: string;
  storyTopic: string;
}

export interface PrepareStoryData {
  topic: string;
  sceneCount: number;
  language: string;
  storyType: HolodeckStoryType;
  wordsPerScene: string;
  selectedCharacters: Character[];
}

export interface MCPServerStatus {
  [serverName: string]: {
    status: 'online' | 'offline';
    latency: number;
  };
}

// System monitoring types
export interface ServiceStatus {
  name: string;
  status: 'online' | 'offline' | 'degraded';
  lastCheck: Date;
  responseTime?: number;
  errorCount?: number;
}

export interface SystemStatus {
  overallHealth: 'healthy' | 'degraded' | 'unhealthy';
  coordinator: ServiceStatus;
  servers: Record<string, ServiceStatus>;
  lastUpdated: Date;
  totalServers: number;
  healthyServers: number;
}

File: software/examples/qollective_mcp_websocket_holodeck/shared-types/src/holodeck.rs
// ABOUTME: Core holodeck types and configurations for Star Trek holodeck experiences
// ABOUTME: Defines the main Holodeck struct and all related types for story management

use serde::{Deserialize, Serialize};
use uuid::Uuid;
use chrono::{DateTime, Utc};
use std::collections::HashMap;
use schemars::JsonSchema;

use crate::characters::Character;
use crate::llm::LlmConfig;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Holodeck {
    pub id: Uuid,
    pub name: String,
    pub topic: String,
    pub story_type: HolodeckStoryType,
    pub participants: Vec<Character>,
    pub current_scene: Option<Scene>,
    pub configuration: HolodeckConfig,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash, JsonSchema)]
pub enum HolodeckStoryType {
    Adventure,
    Mystery,
    Drama,
    Comedy,
    Historical,
    SciFi,
    Fantasy,
    Educational,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Scene {
    pub id: Uuid,
    pub name: String,
    pub description: String,
    pub environment_id: Uuid,
    pub characters_present: Vec<Uuid>,
    pub props: Vec<SceneProp>,
    pub background_audio: Option<String>,
    pub lighting_config: LightingConfig,
    pub physics_settings: PhysicsSettings,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SceneProp {
    pub id: Uuid,
    pub name: String,
    pub description: String,
    pub position: Position3D,
    pub interactive: bool,
    pub physics_enabled: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Position3D {
    pub x: f64,
    pub y: f64,
    pub z: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LightingConfig {
    pub ambient_light: f32,
    pub directional_lights: Vec<DirectionalLight>,
    pub mood: LightingMood,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DirectionalLight {
    pub direction: Position3D,
    pub intensity: f32,
    pub color: Color,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Color {
    pub r: u8,
    pub g: u8,
    pub b: u8,
    pub a: u8,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum LightingMood {
    Bright,
    Dim,
    Dramatic,
    Cozy,
    Mysterious,
    Tense,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PhysicsSettings {
    pub gravity_enabled: bool,
    pub gravity_strength: f32,
    pub collision_detection: bool,
    pub real_time_physics: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HolodeckConfig {
    pub safety_level: SafetyLevel,
    pub max_participants: u32,
    pub duration_minutes: Option<u32>,
    pub auto_save_enabled: bool,
    pub voice_recognition: bool,
    pub haptic_feedback: bool,
    pub replicator_access: bool,
    pub transporter_integration: bool,
    pub environmental_controls: EnvironmentalControls,

    /// Configurable LLM provider settings for character AI and story generation
    pub llm_config: LlmConfig,
}

#[derive(Debug, Clone, Hash, Eq, Serialize, Deserialize, PartialEq, JsonSchema)]
pub enum SafetyLevel {
    Training,    // No harm possible
    Standard,    // Standard safety protocols
    Reduced,     // Some risk allowed
    Disabled,    // All safeties off (rare)
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, JsonSchema)]
pub enum RiskLevel {
    None,        // No risk identified
    Low,         // Minimal risk
    Medium,      // Moderate risk
    High,        // Significant risk
    Critical,    // Severe risk requiring immediate attention
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, JsonSchema)]
pub enum ContentType {
    Educational, // Educational content
    Historical,  // Historical recreations
    Adventure,   // Adventure scenarios
    Drama,       // Dramatic stories
    Comedy,      // Comedy programs
    SciFi,       // Science fiction
    Mystery,     // Mystery scenarios
    Fantasy,     // Fantasy worlds
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EnvironmentalControls {
    pub temperature_celsius: f32,
    pub humidity_percent: f32,
    pub atmospheric_pressure: f32,
    pub oxygen_level: f32,
    pub wind_simulation: bool,
    pub weather_effects: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HolodeckSession {
    pub id: Uuid,
    pub holodeck_id: Uuid,
    pub start_time: DateTime<Utc>,
    pub end_time: Option<DateTime<Utc>>,
    pub participants: Vec<Uuid>,
    pub session_log: Vec<SessionLogEntry>,
    pub status: SessionStatus,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionLogEntry {
    pub timestamp: DateTime<Utc>,
    pub event_type: SessionEventType,
    pub participant_id: Option<Uuid>,
    pub description: String,
    pub metadata: HashMap<String, serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum SessionEventType {
    SessionStart,
    SessionEnd,
    SceneChange,
    CharacterSpawn,
    CharacterDespawn,
    PropInteraction,
    DialogSpoken,
    SafetyProtocolTriggered,
    EnvironmentChange,
    UserCommand,
    SystemEvent,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum SessionStatus {
    Active,
    Paused,
    Completed,
    Aborted,
    SafetyHalt,
}

impl Holodeck {
    pub fn new(name: String, topic: String, story_type: HolodeckStoryType) -> Self {
        let now = Utc::now();
        Self {
            id: Uuid::now_v7(),
            name,
            topic,
            story_type,
            participants: Vec::new(),
            current_scene: None,
            configuration: HolodeckConfig::default(),
            created_at: now,
            updated_at: now,
        }
    }

    pub fn add_participant(&mut self, character: Character) {
        self.participants.push(character);
        self.updated_at = Utc::now();
    }

    pub fn set_scene(&mut self, scene: Scene) {
        self.current_scene = Some(scene);
        self.updated_at = Utc::now();
    }
}

impl Default for HolodeckConfig {
    fn default() -> Self {
        Self {
            safety_level: SafetyLevel::Standard,
            max_participants: 10,
            duration_minutes: None,
            auto_save_enabled: true,
            voice_recognition: true,
            haptic_feedback: true,
            replicator_access: false,
            transporter_integration: false,
            environmental_controls: EnvironmentalControls::default(),
            llm_config: LlmConfig::default(),
        }
    }
}

impl Default for EnvironmentalControls {
    fn default() -> Self {
        Self {
            temperature_celsius: 22.0,
            humidity_percent: 45.0,
            atmospheric_pressure: 101.325,
            oxygen_level: 21.0,
            wind_simulation: false,
            weather_effects: false,
        }
    }
}
